searx/cache.py

```diff
@@ -0,0 +1,461 @@
+
+    cached_items: dict[str, list[tuple[str, typing.Any, int]]]
+    """Values in the cache mapped by table name.
+
+    .. code: python
+
+       {
+           "table name": [
+               ("foo key": "foo value", <expire>),
+               ("bar key": "bar value", <expire>),
+               # ...
+           ],
+           # ...
+       }
+    """
+
+    def report(self):
+        c_tables = 0
+        c_kv = 0
+        lines = []
+
+        for table_name, kv_list in self.cached_items.items():
+            c_tables += 1
+            if not kv_list:
+                lines.append(f"[{table_name:20s}] empty")
+                continue
+
+            for key, value, expire in kv_list:
+                valid_until = datetime.datetime.fromtimestamp(expire).strftime("%Y-%m-%d %H:%M:%S")
+                c_kv += 1
+                lines.append(f"[{table_name:20s}] {valid_until} {key:12}" f" --> ({type(value).__name__}) {value} ")
+
+        lines.append(f"number of tables: {c_tables}")
+        lines.append(f"number of key/value pairs: {c_kv}")
+        return "\n".join(lines)
+
+
+class ExpireCache(abc.ABC):
+    """Abstract base class for the implementation of a key/value cache
+    with expire date."""
+
+    cfg: ExpireCacheCfg
+
+    hmac_iterations: int = 10_000
+    crypt_hash_property = "crypt_hash"
+
+    @abc.abstractmethod
+    def set(self, key: str, value: typing.Any, expire: int | None) -> bool:
+        """Set *key* to *value*.  To set a timeout on key use argument
+        ``expire`` (in sec.).  If expire is unset the default is taken from
+        :py:obj:`ExpireCacheCfg.MAXHOLD_TIME`.  After the timeout has expired,
+        the key will automatically be deleted.
+        """
+
+    @abc.abstractmethod
+    def get(self, key: str, default=None) -> typing.Any:
+        """Return *value* of *key*.  If key is unset, ``None`` is returned."""
+
+    @abc.abstractmethod
+    def maintenance(self, force: bool = False, drop_crypted: bool = False) -> bool:
+        """Performs maintenance on the cache.
+
+        ``force``:
+          Maintenance should be carried out even if the maintenance interval has
+          not yet been reached.
+
+        ``drop_crypted``:
+           The encrypted values can no longer be decrypted (if the password is
+           changed), they must be removed from the cache.
+        """
+
+    @abc.abstractmethod
+    def state(self) -> ExpireCacheStats:
+        """Returns a :py:obj:`ExpireCacheStats`, which provides information
+        about the status of the cache."""
+
+    @staticmethod
+    def build_cache(cfg: ExpireCacheCfg) -> ExpireCache:
+        """Factory to build a caching instance.
+
+        .. note::
+
+           Currently, only the SQLite adapter is available, but other database
+           types could be implemented in the future, e.g. a Valkey (Redis)
+           adapter.
+        """
+        return ExpireCacheSQLite(cfg)
```

Comment:
> As far as I know, SQLite is not really intended to be used as a key value store.

We can implement a Redis/Valkey adapter based on the abstract class `ExpireCache(abc.ABC)`.

> but we should always use Redis if it's available because its performance will be much better as it's actually just made for that single purpose.

I wouldn't expect too much:

1. We only have a few values ​​in the cache (hundreds or less). Key/value databases only show their performance advantages with very large data sets (millions).

2. The SQLite3 DB system runs in-process, meaning the DB query is processed in-process. A key/value DB requires a request/response cycle, which must run over IP stack or at least over sockets.